### ISSUE TITLE
Incorrect return type only detected when value is assigned and returned on separate lines

### ISSUE THREAD

**Issue by Hoolean** [microsoft/pyright#4422]
**Describe the bug**

pyright shows an error when a value of incorrect type is assigned and returned across separate lines, but not when it is returned inline.

**Reproducing code**

```python
# pyright: strict

from __future__ import annotations

from typing import TypeVar

X = TypeVar("X")


def add_to_list_flags(value: X) -> list[X]:
    """Separate assignment and return: does not pass type-check :)"""
    to_add = [value, str(value)]
    return to_add


def add_to_list_silent(value: X) -> list[X]:
    """Inline return: does pass type-check :("""
    return [value, str(value)]


print(add_to_list_flags(5))
print(add_to_list_silent(5))

```

**Expected behavior**

pyright identifies that both functions are returning a value of incorrect type.

**Actual behaviour**

pyright only identifies that `add_to_list_flags` is returning a value of incorrect type; `add_to_list_silent` passes type-checking.

**VS Code extension or command-line**

| | Version |
|-|-|
| VS Code | `1.73.1` |
| Pylance | `v2023.1.10` |

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.289, which I just published. It will also be included in a future release of pylance.
